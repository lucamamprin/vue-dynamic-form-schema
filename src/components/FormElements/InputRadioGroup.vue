<template>
  <div
    v-if="options && name"
    class="input radio"
  >
    <fieldset
      :id="id"
      :aria-required="required.toString()"
      :aria-describedby="describedBy"
      role="radiogroup"
      class="p-0 m-0"
    >
      <legend class="d-flex align-items-center mb-5">
        <span
          v-if="label"
          :class="{'text-primary': !darkMode}"
          class="h6 d-inline-block"
          v-html="label"
        />

        <span v-if="hasError && errorMessage">
          <error-message
            :id="describedBy"
            position="static"
            :inline-error="true"
            class="text-nowrap"
            :error-message="errorMessage"
          />
        </span>
      </legend>

      <div
        v-for="(radio, idx) in options"
        :key="idx"
        class="radio-container-input"
      >
        <input
          :id="getId(idx)"
          type="radio"
          :name="name"
          :value="radio.value"
          :disabled="disabled"
          :aria-invalid="invalid.toString()"
          :aria-labelledby="getId(idx, true)"
          :class="{'missingBackground': hasError && error}"
          @change="$emit('input', radio)"
        >

        <label
          :id="getId(idx, true)"
          :for="getId(idx)"
          :class="{'radio-transparent radio-white text-white': darkMode}"
          v-html="radio.label"
        />
      </div>
    </fieldset>
  </div>
</template>

<script>
import props from "../../mixins/props"
import options from "../../mixins/options"
const ErrorMessage = () => import("./Atoms/ErrorMessage")

export default {
  name: "InputRadioGroup",
  components: {
    ErrorMessage,
  },
  mixins: [
    props,
    options,
  ],
  inheritAttrs: false,
  props: {
    value: {
      type: Object,
      default: () => {},
    },
  },
  methods: {
    getId (idx, isLabel = false) {
      const id = `${this.id}-${idx}`

      return isLabel ? `label-${id}` : id
    },
  },
}
</script>
