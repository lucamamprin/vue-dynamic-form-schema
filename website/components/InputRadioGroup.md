# InputRadioGroup

## Props

<!-- @vuese:InputRadioGroup:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|errorCondition|-|`Boolean`|`false`|-|
|value|-|`Object`|`false`|-|
|id|-|`String`|`true`|-|
|describedBy|-|`String`|`true`|-|
|disabled|-|`Boolean`|`false`|false|
|name|from base select|`String`|`true`|-|
|label|-|`String`|`true`|-|
|options|-|`Array`|`true`|-|
|hasError|-|`Boolean`|`false`|false|
|errorMessage|-|`String`|`false`|null|
|darkMode|-|`Boolean`|`false`|false|
|invalid|-|`Boolean`|`false`|-|
|required|-|`Boolean`|`false`|false|

<!-- @vuese:InputRadioGroup:props:end -->


## Events

<!-- @vuese:InputRadioGroup:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|input|-|-|

<!-- @vuese:InputRadioGroup:events:end -->


