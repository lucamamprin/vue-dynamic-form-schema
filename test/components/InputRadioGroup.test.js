import Vue from "vue"
import { shallowMount } from "@vue/test-utils"
import InputRadioGroup from "../../src/components/FormElements/InputRadioGroup.vue"

const props = {
  name: "hobby",
  id: "user-hobby",
  label: "What is your hobby?",
  describedBy: "hobby-describe",
  options: [
    {
      "label": "Cinema",
      "value": "cinema",
    },
    {
      "label": "Bird watching",
      "value": "bird_watching",
    },
    {
      "label": "Coding",
      "value": "coding",
    },
  ],
}

let wrapper

describe("InputRadioGroup", () => {
  beforeEach(() => {
    wrapper = shallowMount(InputRadioGroup, {
      propsData: props,
    })
  })
  
  test("mounts correctly", () => {
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })
  
  test("validates options", () => {
    const validator = InputRadioGroup.props.options.validator
    expect(validator(props.options)).toBe(true)
  })
  
  test("validates wrong options", () => {
    const wrongOptions =  [
      {
        "wrong-label": "Cinema",
        "wrong-value": "cinema",
      },
    ]

    const validator = InputRadioGroup.props.options.validator
    expect(validator(wrongOptions)).toBe(false)
  })
  
  test("can be disabled", async () => {
    await wrapper.setProps({
      disabled: true,
    })

    await Vue.nextTick()
    
    expect(wrapper.html()).toMatchSnapshot()
  })
  
  test("shows error message", async () => {
    await wrapper.setProps({
      hasError: true,
      errorMessage: "You didn't select any hobby",
    })

    await Vue.nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })
})