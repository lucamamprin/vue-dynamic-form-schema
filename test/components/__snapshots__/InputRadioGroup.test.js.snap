// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRadioGroup can be disabled 1`] = `
<div class="input radio">
  <fieldset id="user-hobby" aria-required="false" aria-describedby="hobby-describe" role="radiogroup" class="p-0 m-0">
    <legend class="d-flex align-items-center mb-5"><span class="h6 d-inline-block text-primary">What is your hobby?</span>
      <!---->
    </legend>
    <div class="radio-container-input"><input id="user-hobby-0" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-0" class="" value="cinema" disabled="disabled"> <label id="label-user-hobby-0" for="user-hobby-0" class="">Cinema</label></div>
    <div class="radio-container-input"><input id="user-hobby-1" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-1" class="" value="bird_watching" disabled="disabled"> <label id="label-user-hobby-1" for="user-hobby-1" class="">Bird watching</label></div>
    <div class="radio-container-input"><input id="user-hobby-2" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-2" class="" value="coding" disabled="disabled"> <label id="label-user-hobby-2" for="user-hobby-2" class="">Coding</label></div>
  </fieldset>
</div>
`;

exports[`InputRadioGroup mounts correctly 1`] = `
<div class="input radio">
  <fieldset id="user-hobby" aria-required="false" aria-describedby="hobby-describe" role="radiogroup" class="p-0 m-0">
    <legend class="d-flex align-items-center mb-5"><span class="h6 d-inline-block text-primary">What is your hobby?</span>
      <!---->
    </legend>
    <div class="radio-container-input"><input id="user-hobby-0" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-0" class="" value="cinema"> <label id="label-user-hobby-0" for="user-hobby-0" class="">Cinema</label></div>
    <div class="radio-container-input"><input id="user-hobby-1" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-1" class="" value="bird_watching"> <label id="label-user-hobby-1" for="user-hobby-1" class="">Bird watching</label></div>
    <div class="radio-container-input"><input id="user-hobby-2" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-2" class="" value="coding"> <label id="label-user-hobby-2" for="user-hobby-2" class="">Coding</label></div>
  </fieldset>
</div>
`;

exports[`InputRadioGroup shows error message 1`] = `
<div class="input radio">
  <fieldset id="user-hobby" aria-required="false" aria-describedby="hobby-describe" role="radiogroup" class="p-0 m-0">
    <legend class="d-flex align-items-center mb-5"><span class="h6 d-inline-block text-primary">What is your hobby?</span> <span><span id="hobby-describe" class="help m-0 text-nowrap ml-1 position-static d-inline-block">
  You didn't select any hobby
</span></span></legend>
    <div class="radio-container-input"><input id="user-hobby-0" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-0" class="" value="cinema"> <label id="label-user-hobby-0" for="user-hobby-0" class="">Cinema</label></div>
    <div class="radio-container-input"><input id="user-hobby-1" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-1" class="" value="bird_watching"> <label id="label-user-hobby-1" for="user-hobby-1" class="">Bird watching</label></div>
    <div class="radio-container-input"><input id="user-hobby-2" type="radio" name="hobby" aria-invalid="false" aria-labelledby="label-user-hobby-2" class="" value="coding"> <label id="label-user-hobby-2" for="user-hobby-2" class="">Coding</label></div>
  </fieldset>
</div>
`;
